### PR TITLE
Add commutator functions to quantum_info

### PR DESCRIPTION
**Pull request by ikkoham**
### Summary

Add commutator functions to quantum_info.

Note that unlike opflow, it does not perform `reduce` (simplify) method. This is because `LinearOp` is not guaranteed to have a simplify method. Therefore, when calling it, users must explicitly simplify after the call if necessary.

### CODE REVIEW

File: test/python/quantum_info/operators/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+"""Tests utility functions for operator classes."""
+
+import unittest
+
+from qiskit.quantum_info import SparsePauliOp, anti_commutator, commutator, double_commutator
+from qiskit.test import QiskitTestCase
+
+
+class TestOperatorUtils(QiskitTestCase):
```
Comment (mrossinek) [inline test/python/quantum_info/operators/test_utils.py:21]: It might be worthwhile adding one additional test in each case to ensure that the method does not _always_ return zero-ops. Because if this were to break in the future, these tests would not catch that.

**Review by mrossinek — COMMENTED** [review]
Just two minor observations but in general this LGTM :+1:

File: qiskit/quantum_info/operators/utils.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    Returns:
+        The anti-commutator
+    """
+    return a @ b + b @ a
+
+
+def double_commutator(a: LinearOp, b: LinearOp, c: LinearOp, *, sign: bool = False) -> LinearOp:
+    r"""
+    Compute symmetric double commutator of a, b and c.
+    See McWeeny chapter 13.6 Equation of motion methods (page 479)
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:57]: With a single header line followed by an empty line this renders nicely in the docs 🙂 
```suggestion
    r"""Compute symmetric double commutator of a, b and c.

    See McWeeny chapter 13.6 Equation of motion methods (page 479)
```

File: qiskit/quantum_info/operators/utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    .. math::
+         \lbrace[A, B], C\rbrace/2 + \lbrace A, [B, C]\rbrace/2
+         = (2ABC - 2CBA - BAC + CAB - ACB + BCA)/2.
+
+    Args:
+        a: Operator a.
+        b: Operator b.
+        c: Operator c.
+        sign: False anti-commutes, True commutes.
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:76]: If this argument specifies whether a commutator or anticommutator is computed, how about reflecting this in the argument and rename it to ``anticommutator`` (``False`` per default)?

**Comment by ikkoham** [comment]
Thank you very much for great suggestions. I improved codes.

**Review by mrossinek — DISMISSED** [review]
LGTM :+1:

File: qiskit/quantum_info/operators/utils.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+"""
+Quantum information utility functions for operators.
+"""
+
+from typing import TypeVar
+
+from qiskit.quantum_info.operators.linear_op import LinearOp
+
+OperatorTypeT = TypeVar("OperatorTypeT", bound=LinearOp)
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:21]: This is a general question as I'm not that familiar with `TypeVar`s 🙂 What's the advantage of using a type var with a `bound` argument over just the `LinearOp` itself? I.e. in your code, what would be the difference to writing
```python
def commutator(a: LinearOp, b: LinearOp) -> LinearOp
```
?

File: qiskit/quantum_info/operators/utils.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    .. math::
+
+         [[A, B], C]/2 + [A, [B, C]]/2
+         = (2ABC + 2CBA - BAC - CAB - ACB - BCA)/2.
+
+    If `commutes` is `False`, it returns
+
+    .. math::
+         \lbrace[A, B], C\rbrace/2 + \lbrace A, [B, C]\rbrace/2
+         = (2ABC - 2CBA - BAC + CAB - ACB + BCA)/2.
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:73]: Does this render correctly in the docs or does this need another linebreak?
```suggestion
    .. math::

         \lbrace[A, B], C\rbrace/2 + \lbrace A, [B, C]\rbrace/2
         = (2ABC - 2CBA - BAC + CAB - ACB + BCA)/2.
```

File: qiskit/quantum_info/operators/utils.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        The anti-commutator
+    """
+    return a @ b + b @ a
+
+
+def double_commutator(
+    a: OperatorTypeT, b: OperatorTypeT, c: OperatorTypeT, *, commutes: bool = True
+) -> OperatorTypeT:
+    r"""Compute symmetric double commutator of a, b and c.
+    See also Equation (13.6.18) in [1].
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:60]: I think this needs a line break to make sure the first line is recognized as summary.
```suggestion
    r"""Compute symmetric double commutator of a, b and c.

    See also Equation (13.6.18) in [1].
```

File: qiskit/quantum_info/operators/utils.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    .. math::
+         \lbrace[A, B], C\rbrace/2 + \lbrace A, [B, C]\rbrace/2
+         = (2ABC - 2CBA - BAC + CAB - ACB + BCA)/2.
+
+    Args:
+        a: Operator a.
+        b: Operator b.
+        c: Operator c.
+        commutes: True commutes, False anti-commutes.
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:79]:
```suggestion
        commutes: ``True`` commutes, ``False`` anti-commutes.
```

File: qiskit/quantum_info/operators/utils.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    Returns:
+        The double commutator
+
+    References:
+        [1]: R. McWeeny.
+            Methods of Molecular Quantum Mechanics.
+            2nd Edition, Academic Press, 1992.
+            ISBN 0-12-486552-6.
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:88]: The indents here are off by a space, does that cause any weird formatting in the docs?

File: releasenotes/notes/add-commutator-96ef07433e8ca4e7.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Add utility functions :func:`~qiskit.quantum_info.commutator`,
+    :func:`~qiskit.quantum_info.anti_commutator`, and
+    :func:`~qiskit.quantum_info.double_commutator`.
```
Comment (Cryoris) [inline releasenotes/notes/add-commutator-96ef07433e8ca4e7.yaml:6]:
```suggestion
    :func:`~qiskit.quantum_info.double_commutator` to compute commutators
    for any objects implementing a :class:`.LinearOp`.
```

File: qiskit/quantum_info/operators/utils.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    See also Equation (13.6.18) in [1].
+
+    If `commutes` is `True`, it returns
+
+    .. math::
+
+         [[A, B], C]/2 + [A, [B, C]]/2
+         = (2ABC + 2CBA - BAC - CAB - ACB - BCA)/2.
+
+    If `commutes` is `False`, it returns
```
Comment (Cryoris) [inline qiskit/quantum_info/operators/utils.py:69]: As an adjective "commutes" sounds like a property of the operators, but here this decides whether a commutator or anticommutator is computed. I think calling this either `commutator=True` or `anticommutator=False` would be clearer 🙂

**Review by Cryoris — COMMENTED** [review]
Some minor comments and one question, otherwise LGTM too!

File: qiskit/quantum_info/operators/utils.py (hunk 4 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/utils.py:21]: There is  clear advantage for `TypeVar`. If a and b are `SparsePauliOp` the return type should be `SparsePauliOp`, but your type hint just said `LinearOp`..

File: qiskit/quantum_info/operators/utils.py (hunk 10 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/utils.py:69]: I'd prefer that one, but I can't use it because it conflicts with the function name. I came up with nice idea. I will separate the files.

File: qiskit/quantum_info/operators/utils.py (hunk 8 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/utils.py:88]: No. Probably not for any deep reason, but because it's easy to read. (These lines come from nature)

File: qiskit/quantum_info/operators/utils.py (hunk 5 above)
Comment (ikkoham) [reply qiskit/quantum_info/operators/utils.py:73]: I'm looking at the rendering results, but probably not necessary.

image

File: qiskit/quantum_info/operators/utils.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    Args:
+        a: Operator a.
+        b: Operator b.
+        c: Operator c.
+        commutes: True commutes, False anti-commutes.
+
+    Returns:
+        The double commutator
+
+    References:
```
Comment (ikkoham) [inline qiskit/quantum_info/operators/utils.py:84]:
```suggestion
    References:

```

File: qiskit/quantum_info/operators/utils.py (hunk 4 above)
Comment (Cryoris) [reply qiskit/quantum_info/operators/utils.py]: Ah, I see! Thanks for the explanation 👍🏻

**Review by Cryoris — DISMISSED** [review]
LGTM thanks for integrating the changes!

**Review by Cryoris — APPROVED** [review]
